### PR TITLE
Convert to using WEAK/STRONG macros

### PR DESCRIPTION
**Pull request by bdorfman**
These macros are based off libextc's @weakify and @strongify.
This change also fixes some incorrect self-capturing retain cycles and probably some potential errors with weakself being dealloc'd partway through block execution.

### CODE REVIEW

File: Stripe/STPCheckoutAPIClient.m [hunk 1]
```diff
@@ -77,27 +78,28 @@ - (BOOL)readyForLookups {
 }
 
 - (STPPromise *)lookupEmail:(NSString *)email {
-    __weak typeof(self) weakself = self;
+    WEAK(self);
+    Class selfClass = self.class;
```
Comment (benzguo) [inline Stripe/STPCheckoutAPIClient.m]: curious – why is this necessary?

File: Stripe/STPPaymentContext.m [hunk 2]
```diff
@@ -69,29 +70,32 @@ - (void)retryLoading {
...
-                    [weakself.delegate paymentContext:weakself didFailToLoadWithError:error];
+                    [self.delegate paymentContext:self didFailToLoadWithError:error];
                 }
             }];
         }
     }];
     [self.apiAdapter retrieveCustomer:^(STPCustomer * _Nullable customer, NSError * _Nullable error) {
-        if (!weakself) {
+        STRONG(self);
+        if (!self) {
```
Comment (benzguo) [inline Stripe/STPPaymentContext.m]: it's odd that we weren't strongifying self in any callbacks in this file...

**Comment by benzguo** [comment]
LGTM

File: Stripe/STPCheckoutAPIClient.m (hunk 1 above)
Comment (bdorfman) [inline Stripe/STPCheckoutAPIClient.m]: So if you pull out selfClass here, then you are capturing `selfClass`. If you reference `self.class` later, that is capturing `self` (and being converted to `[self class]` basically). Under the previous code, the weak self can never actually be nil because you are capturing it inside the `if (!weakself) {}` block. In the new method, the if statement works correctly, plus we still have a ref to the class type for use in the error message.
